### DISCUSSION TITLE
Space Shuttle Software

### DISCUSSION THREAD

**Discussion by ColanderCombo** [virtualagc/virtualagc#1219] [General]
Howdy!

I'm excited to see that you managed to find a HAL/S compiler and the work you've been doing to resurrect it in some form.  I've been doing my own work collecting documentation on shuttle hardware and software, and made some attempts at emulation of the GPC and MEDS system.  I've spent some time rebuilding portions of FCOS and PASS from available specifications.  For example, here's my current version of SM2_OPS (OPS 2 main loop) and SSB_PL_BAY_DOORS (Payload Bay Doors SPEC function):

https://gist.github.com/ColanderCombo/3b28cc7c934c801ee5b37db8152a01a6

I just found your work yesterday evening and haven't had a chance to run it through your compiler, but it does parse successfully in my own HAL/S grammar.

I've also been working on rebuilding some version of the CCMS to handle ground systems.  The specs are 30k page so things are progressing slowly, but if people here are interested I can share more.

I'd love to help out in your effort in any way that might be useful.  This stuff is fantastic!

**Comment by rburkey2005** [comment]
@ColanderCombo Very interesting indeed!

Just to clarify ... from the way you phrased it, I gather you have some way to parse the HAL/S code you've written so that you can determine that it's correct syntactically, but don't have a compiler or some other means to execute the code?

What _I_ have is two compilers, neither of which is complete at present.  

One (which I call the "modern compiler") was written from scratch, but it does not yet support some of the fancier HAL/S features.  However, it does compile all the way to a form that I call PALMAT, and I have a PALMAT emulator that can run the compiled code, albeit very slowly.  The compiler and emulator can be used together, to effectively create a HAL/S interactive interpreter.

The other (the "original compiler") is a port of the Intermetrics HAL/S compiler's XPL source code to Python.  It supports the entire HAL/S feature set.  However, only the 1st pass (of 7) has been ported, so what it compiles to is the so-called HALMAT "intermediate language".  There is no HALMAT emulator, and the documentation for HALMAT has mostly not survived.  In principle, if several more passes of the original compiler are ported, then it could compile all the way to GPC assembly language or to IBM/360 assembly language.

And yes, I need help &mdash; lots of help &mdash; particularly with continuing the original compiler, because it's very time-consuming work and I've burned myself out on it.  The basic structure of the original compiler is this:

1. Pass 1 creates HALMAT "intermediate language" from HAL/S source code.  (This is complete, but only marginally tested.)
2. Three optimizing passes, but only two of which actually modify the HALMAT.  (Unfortunately, optimization seems to be required rather than optional.)
3. Pass 2 creates GPC or System/360 object code from optimized HALMAT.
4. Pass 3 and Pass 4 which modify neither HALMAT, GPC, nor System/360 object code, and are therefore not needed.

So the help I _really_ need is porting the two necessary optimizing passes and Pass 2. 

Alternately, figuring out some way for Pass 1 to output unoptimized but executable code (say, in C) would be nice.

Failing that, there are some other possibilities, such as testing the HALMAT generation from Pass 1, or working out some way of emulating execution of HALMAT; but given that only about 20% of the HALMAT documentation is available, that's likely to be pretty tough.

**Reply by ColanderCombo** [reply]
parser: yeah, I have a parser based on the PASS 32.0/BFS 17.0 HAL/S Language Specification (USA-003088).  It can parse every scrap of HAL/S I've come across or written and It can dump some *very* partial BAL code, but is nowhere near complete enough to be useful.  I also have a naive BAL assembler that can handle all the AP-101 instructions but is missing macros and other important features.  I can run the assembled  object code on a high-level AP-101 simulator. 

I should stress that all of this is held together with duct tape and chewing gum.  The AP-101 simulator, for example, stores floats in AP-101 format but does all the floating point operations using javascript floats.  My main goal was to get some basic end-to-end understanding of what was going on and build from there.

Having an even partially complete modern compiler is an impressive achievement! I'm taking a look at this and will see if I can get any of my code to run under your interpreter!

I've looked through the available docs on the 360 compiler, and I think I found the same original XPL source you have in your tree. I think we agree that it would be amazing to get the original compiler compiled from original source, but there isn't enough there to do that right now.

I'm super stoked to take a crack at the optimizer and Pass 2.  If we can get to BAL or object code I believe there's a direct path to running it under an emulator.  

I've been focusing on CCMS for a while so my GPC emulator is currently pretty crufty.  I'm pulling it back up now to see what state it's in.  It's currently written in coffeescript, which was a poor decision.   Rewriting in something more sensible like python is on my list.

I'm gonna dig into the original compiler source and your port  and see what's going on!

**Comment by ColanderCombo** [comment]
@rburkey2005 OMG you have the entire FSSR!  I've been looking for these for YEARS, but only managed to find portions.  If it's useful, I do have portions of the OI-8B version that I scanned from paper.  Your  cache is extensive, but I may also have some other docs that may be of interest.

**Reply by rburkey2005** [reply]
Anything missing from the library is welcome.  Admittedly, I do have a cache of stuff that has been sent to me that hasn't been uploaded to the library as of yet.

**Reply by ColanderCombo** [reply]
Cool! Is there a preferred way to get this to you?  Mail to your address on the website bounces.

**Reply by rburkey2005** [reply]
Tricky.  

I find I can't email to that address either, though there's no indication for me of a bounce; my emails simply never arrive.  It appears not to have been working for over a month, and I simply didn't notice.  We could just wait until that's fixed, or we could look for some interim solution.

Meanwhile, github used to have a feature for sending a private message to another user, but seems to have discontinued it, so I can't use it to just send you a different email address.  And I don't want to publish my main email address on a public site.

Do you have a google drive or something similar you can temporarily post the scan to, and put the link to it here?

**Reply by rburkey2005** [reply]
@ColanderCombo The email address given on the website now works.  I don't know how large its file attachments can be.  If it rejects the file attachment, I'd suggest just sending me a message without an attachment, and I'll reply from a better email address.

**Reply by ColanderCombo** [reply]
Awesome!  I just sent off a note with a link to a shared folder.

**Comment by rburkey2005** [comment]
> I've looked through the available docs on the 360 compiler, and I think I found the same original XPL source you have in your tree. I think we agree that it would be amazing to get the original compiler compiled from original source, but there isn't enough there to do that right now.

I _think_ what you're referring to here is the Intermetrics XPL compiler that would have been used to compile the Intermetrics HAL/S compiler.  You're correct if you're saying that we don't have the source code for that XPL compiler; as far as I know, it has not survived, and I see no possibility of resurrecting it from what little material remains.  The XPL compiler in the source tree is a "standard" XPL compiler that doesn't include Intermetrics's enhancements.  It cannot compile the Intermetrics HAL/S compiler, and if I knew then what I know now, I wouldn't have bothered to put it in the source tree.  :smiley:

**Reply by ColanderCombo** [reply]
Yeah, I'm referring to the original non-Intermetrics XPL.  I came to the same conclusion as you.

**Comment by rburkey2005** [comment]
> I should stress that all of this is held together with duct tape and chewing gum.

That's basically what I'd say about the "modern" HAL/S compiler, which is the reason I (temporarily) stopped working on it to port the "original" compiler instead.  (I found that the HAL/S grammar wasn't context-free, but that I could fake up a context-free grammar with a preprocessor and a bunch of new BNF grammar rules.  With all that junk bolted together, the modern compiler is quite hard to maintain and extend.)

Meanwhile, I found that porting the original compiler is quite tricky too, because there's a bunch of undocumented trickery going on that relies on the specific behavior of the system/360 computer the compiler was intended to run on.  So porting the compiler is quite the adventure, and it _too_ ends up being held together with duct tape.
